<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.32">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Julien Fouret">

<title>Primary Analysis: Quality Control and Preprocessing â€“ De Novo Genome Assembly Tutorial</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-37eea08aefeeee20ff55810ff984fec1.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap-3fced2653c1411365395c6308d7f7025.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="../styles.css">
</head>

<body class="floating nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">De Novo Genome Assembly Tutorial</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-primary-analysis" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Primary Analysis</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-primary-analysis">    
        <li>
    <a class="dropdown-item" href="../primary-analysis/base-calling-qc.html">
 <span class="dropdown-text">Base Calling &amp; QC</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../primary-analysis/preliminary-analysis.html">
 <span class="dropdown-text">Preliminary Analysis</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-hands-on-practice" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Hands-on Practice</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-hands-on-practice">    
        <li>
    <a class="dropdown-item" href="../hands-on/01-fetch-public-datasets.html">
 <span class="dropdown-text">Fetch Public Datasets</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../hands-on/02-primary-analysis.html">
 <span class="dropdown-text">Primary analysis</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../hands-on/03-kmer-analysis.html">
 <span class="dropdown-text">Kmer-based Analysis</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../hands-on/04-hybrid-assembly-spades-masurca.html">
 <span class="dropdown-text">Hybrid Assembly (SPAdes/MaSuRCA)</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../hands-on/05-assembly-evaluation.html">
 <span class="dropdown-text">Assembly Evaluation</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../hands-on/06-scaffolding.html">
 <span class="dropdown-text">Secondary scaffolding</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../hands-on/07-repeats-analysis.html">
 <span class="dropdown-text">Repeats Annotation and Masking</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../hands-on/08-prepare-data-for-annotation.html">
 <span class="dropdown-text">Prepare data for Annotation</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../hands-on/09-annot-with-braker.html">
 <span class="dropdown-text">Gene Annotation with BRAKER</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../hands-on/10-annot-functional.html">
 <span class="dropdown-text">Functional Gene Annotation</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../hands-on/11-comparative-genomics.html">
 <span class="dropdown-text">Comparative Genomics</span></a>
  </li>  
    </ul>
  </li>
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../about.html"> 
<span class="menu-text">About</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#primary-analysis-quality-control-and-preprocessing" id="toc-primary-analysis-quality-control-and-preprocessing" class="nav-link active" data-scroll-target="#primary-analysis-quality-control-and-preprocessing"><span class="header-section-number">2</span> Primary Analysis: Quality Control and Preprocessing</a>
  <ul class="collapse">
  <li><a href="#configuration-and-setup" id="toc-configuration-and-setup" class="nav-link" data-scroll-target="#configuration-and-setup"><span class="header-section-number">2.1</span> Configuration and Setup</a>
  <ul class="collapse">
  <li><a href="#environment-variables" id="toc-environment-variables" class="nav-link" data-scroll-target="#environment-variables"><span class="header-section-number">2.1.1</span> Environment Variables</a></li>
  <li><a href="#software-dependencies" id="toc-software-dependencies" class="nav-link" data-scroll-target="#software-dependencies"><span class="header-section-number">2.1.2</span> Software Dependencies</a></li>
  </ul></li>
  <li><a href="#directory-structure-setup" id="toc-directory-structure-setup" class="nav-link" data-scroll-target="#directory-structure-setup"><span class="header-section-number">2.2</span> Directory Structure Setup</a></li>
  <li><a href="#initial-quality-control-assessment" id="toc-initial-quality-control-assessment" class="nav-link" data-scroll-target="#initial-quality-control-assessment"><span class="header-section-number">2.3</span> Initial Quality Control Assessment</a>
  <ul class="collapse">
  <li><a href="#illumina-data-qc-with-fastqc" id="toc-illumina-data-qc-with-fastqc" class="nav-link" data-scroll-target="#illumina-data-qc-with-fastqc"><span class="header-section-number">2.3.1</span> Illumina Data QC with FastQC</a></li>
  <li><a href="#nanopore-data-qc-with-nanoplot" id="toc-nanopore-data-qc-with-nanoplot" class="nav-link" data-scroll-target="#nanopore-data-qc-with-nanoplot"><span class="header-section-number">2.3.2</span> Nanopore Data QC with NanoPlot</a></li>
  </ul></li>
  <li><a href="#preprocessing-filtering-and-trimming" id="toc-preprocessing-filtering-and-trimming" class="nav-link" data-scroll-target="#preprocessing-filtering-and-trimming"><span class="header-section-number">2.4</span> Preprocessing: Filtering and Trimming</a>
  <ul class="collapse">
  <li><a href="#illumina-data-preprocessing-with-fastp" id="toc-illumina-data-preprocessing-with-fastp" class="nav-link" data-scroll-target="#illumina-data-preprocessing-with-fastp"><span class="header-section-number">2.4.1</span> Illumina Data Preprocessing with fastp</a></li>
  <li><a href="#nanopore-data-preprocessing" id="toc-nanopore-data-preprocessing" class="nav-link" data-scroll-target="#nanopore-data-preprocessing"><span class="header-section-number">2.4.2</span> Nanopore Data Preprocessing</a></li>
  </ul></li>
  <li><a href="#read-merging-illumina-only" id="toc-read-merging-illumina-only" class="nav-link" data-scroll-target="#read-merging-illumina-only"><span class="header-section-number">2.5</span> Read Merging (Illumina Only)</a></li>
  <li><a href="#post-processing-quality-control" id="toc-post-processing-quality-control" class="nav-link" data-scroll-target="#post-processing-quality-control"><span class="header-section-number">2.6</span> Post-Processing Quality Control</a>
  <ul class="collapse">
  <li><a href="#post-processing-qc-for-illumina-data" id="toc-post-processing-qc-for-illumina-data" class="nav-link" data-scroll-target="#post-processing-qc-for-illumina-data"><span class="header-section-number">2.6.1</span> Post-Processing QC for Illumina Data</a></li>
  <li><a href="#post-processing-qc-for-nanopore-data" id="toc-post-processing-qc-for-nanopore-data" class="nav-link" data-scroll-target="#post-processing-qc-for-nanopore-data"><span class="header-section-number">2.6.2</span> Post-Processing QC for Nanopore Data</a></li>
  </ul></li>
  <li><a href="#taxonomic-classification-with-kraken2" id="toc-taxonomic-classification-with-kraken2" class="nav-link" data-scroll-target="#taxonomic-classification-with-kraken2"><span class="header-section-number">2.7</span> Taxonomic Classification with Kraken2</a>
  <ul class="collapse">
  <li><a href="#database-setup" id="toc-database-setup" class="nav-link" data-scroll-target="#database-setup"><span class="header-section-number">2.7.1</span> Database Setup</a></li>
  <li><a href="#subsampling-with-seqtk" id="toc-subsampling-with-seqtk" class="nav-link" data-scroll-target="#subsampling-with-seqtk"><span class="header-section-number">2.7.2</span> Subsampling with seqtk</a></li>
  <li><a href="#kraken2-classification" id="toc-kraken2-classification" class="nav-link" data-scroll-target="#kraken2-classification"><span class="header-section-number">2.7.3</span> Kraken2 Classification</a></li>
  <li><a href="#interpreting-kraken2-results" id="toc-interpreting-kraken2-results" class="nav-link" data-scroll-target="#interpreting-kraken2-results"><span class="header-section-number">2.7.4</span> Interpreting Kraken2 Results</a></li>
  </ul></li>
  <li><a href="#comprehensive-reporting-with-multiqc" id="toc-comprehensive-reporting-with-multiqc" class="nav-link" data-scroll-target="#comprehensive-reporting-with-multiqc"><span class="header-section-number">2.8</span> Comprehensive Reporting with MultiQC</a></li>
  <li><a href="#expected-outcomes" id="toc-expected-outcomes" class="nav-link" data-scroll-target="#expected-outcomes"><span class="header-section-number">2.9</span> Expected Outcomes</a>
  <ul class="collapse">
  <li><a href="#illumina-data-improvements" id="toc-illumina-data-improvements" class="nav-link" data-scroll-target="#illumina-data-improvements"><span class="header-section-number">2.9.1</span> Illumina Data Improvements:</a></li>
  <li><a href="#nanopore-data-improvements" id="toc-nanopore-data-improvements" class="nav-link" data-scroll-target="#nanopore-data-improvements"><span class="header-section-number">2.9.2</span> Nanopore Data Improvements:</a></li>
  </ul></li>
  <li><a href="#next-steps" id="toc-next-steps" class="nav-link" data-scroll-target="#next-steps"><span class="header-section-number">2.10</span> Next Steps</a></li>
  <li><a href="#multiple-datasets-for-assembly" id="toc-multiple-datasets-for-assembly" class="nav-link" data-scroll-target="#multiple-datasets-for-assembly"><span class="header-section-number">2.11</span> Multiple datasets for assembly</a>
  <ul class="collapse">
  <li><a href="#the-quality-vs.-depth-trade-off" id="toc-the-quality-vs.-depth-trade-off" class="nav-link" data-scroll-target="#the-quality-vs.-depth-trade-off"><span class="header-section-number">2.11.1</span> The Quality vs.&nbsp;Depth Trade-off</a></li>
  <li><a href="#sequencing-depth-analysis" id="toc-sequencing-depth-analysis" class="nav-link" data-scroll-target="#sequencing-depth-analysis"><span class="header-section-number">2.11.2</span> Sequencing depth analysis</a></li>
  </ul></li>
  </ul></li>
  </ul>
</nav>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar zindex-bottom">
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Primary Analysis: Quality Control and Preprocessing</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Julien Fouret </p>
          </div>
  </div>
    
  
    
  </div>
  


</header>


<section id="primary-analysis-quality-control-and-preprocessing" class="level1" data-number="2">
<h1 data-number="2"><span class="header-section-number">2</span> Primary Analysis: Quality Control and Preprocessing</h1>
<p>After downloading our genomic datasets, the next critical step is to assess data quality and perform necessary preprocessing. This chapter covers quality control (QC) assessment, read filtering, trimming, and merging operations for both Illumina short-read and Oxford Nanopore long-read data.</p>
<p>Quality control is essential because: - Raw sequencing data often contains low-quality bases, especially at read ends - Adapter sequences may remain attached to reads - Some reads may be too short to be useful for assembly - Understanding data characteristics helps optimize downstream analysis parameters</p>
<p>Weâ€™ll use a comprehensive workflow that includes:</p>
<ol type="1">
<li><strong>Initial QC assessment</strong> to understand raw data characteristics</li>
<li><strong>Preprocessing</strong> (filtering, trimming, merging) to improve data quality</li>
<li><strong>Post-processing QC</strong> to verify improvements</li>
<li><strong>Comprehensive reporting</strong> with MultiQC</li>
</ol>
<section id="configuration-and-setup" class="level2" data-number="2.1">
<h2 data-number="2.1" class="anchored" data-anchor-id="configuration-and-setup"><span class="header-section-number">2.1</span> Configuration and Setup</h2>
<section id="environment-variables" class="level3" data-number="2.1.1">
<h3 data-number="2.1.1" class="anchored" data-anchor-id="environment-variables"><span class="header-section-number">2.1.1</span> Environment Variables</h3>
<p>Weâ€™ll continue using the environment variables from the previous chapter. If starting fresh, set them again:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="bu">export</span> <span class="va">SN</span><span class="op">=</span>tuto482</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="bu">export</span> <span class="va">NCPUS</span><span class="op">=</span>12</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="software-dependencies" class="level3" data-number="2.1.2">
<h3 data-number="2.1.2" class="anchored" data-anchor-id="software-dependencies"><span class="header-section-number">2.1.2</span> Software Dependencies</h3>
<p>This workflow uses several bioinformatics tools, each optimized for specific data types and tasks:</p>
<ul>
<li><strong>FastQC</strong>: Quality assessment for Illumina short reads</li>
<li><strong>NanoPlot</strong>: Quality assessment specialized for long reads (Nanopore/PacBio)</li>
<li><strong>fastp</strong>: Ultra-fast preprocessing tool for both short and long reads</li>
<li><strong>MultiQC</strong>: Aggregates QC reports from multiple tools</li>
</ul>
<section id="apptainer-aliases" class="level4" data-number="2.1.2.1">
<h4 data-number="2.1.2.1" class="anchored" data-anchor-id="apptainer-aliases"><span class="header-section-number">2.1.2.1</span> Apptainer Aliases</h4>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="bu">alias</span> fastqc=<span class="st">"apptainer run docker://staphb/fastqc fastqc"</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="bu">alias</span> NanoPlot=<span class="st">"apptainer run docker://staphb/nanoplot NanoPlot"</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="bu">alias</span> fastp=<span class="st">"apptainer run docker://staphb/fastp fastp"</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="bu">alias</span> multiqc=<span class="st">"apptainer run docker://staphb/multiqc multiqc"</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="bu">alias</span> seqtk=<span class="st">"apptainer run docker://staphb/seqtk seqtk"</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="bu">alias</span> kraken2=<span class="st">"apptainer run docker://staphb/kraken2 kraken2"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="docker-aliases" class="level4" data-number="2.1.2.2">
<h4 data-number="2.1.2.2" class="anchored" data-anchor-id="docker-aliases"><span class="header-section-number">2.1.2.2</span> Docker Aliases</h4>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="bu">alias</span> fastqc=<span class="st">"docker run --rm -u </span><span class="va">$UID</span><span class="st">:</span><span class="va">$GID</span><span class="st"> -v </span><span class="dt">\$</span><span class="st">(pwd):/data -w /data staphb/fastqc fastqc"</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="bu">alias</span> nanoplot=<span class="st">"docker run --rm -u </span><span class="va">$UID</span><span class="st">:</span><span class="va">$GID</span><span class="st"> -v </span><span class="dt">\$</span><span class="st">(pwd):/data -w /data staphb/nanoplot NanoPlot"</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="bu">alias</span> fastp=<span class="st">"docker run --rm -u </span><span class="va">$UID</span><span class="st">:</span><span class="va">$GID</span><span class="st"> -v </span><span class="dt">\$</span><span class="st">(pwd):/data -w /data staphb/fastp fastp"</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="bu">alias</span> multiqc=<span class="st">"docker run --rm -u </span><span class="va">$UID</span><span class="st">:</span><span class="va">$GID</span><span class="st"> -v </span><span class="dt">\$</span><span class="st">(pwd):/data -w /data staphb/multiqc multiqc"</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="bu">alias</span> seqtk=<span class="st">"docker run --rm -u </span><span class="va">$UID</span><span class="st">:</span><span class="va">$GID</span><span class="st"> -v </span><span class="dt">\$</span><span class="st">(pwd):/data -w /data staphb/seqtk seqtk"</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="bu">alias</span> kraken2=<span class="st">"docker run --rm -u </span><span class="va">$UID</span><span class="st">:</span><span class="va">$GID</span><span class="st"> -v </span><span class="dt">\$</span><span class="st">(pwd):/data -w /data staphb/kraken2 kraken2"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="conda" class="level4" data-number="2.1.2.3">
<h4 data-number="2.1.2.3" class="anchored" data-anchor-id="conda"><span class="header-section-number">2.1.2.3</span> Conda</h4>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create environment with all required tools</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> create <span class="at">-n</span> qc-tools <span class="at">-c</span> bioconda fastqc nanoplot fastp kraken2 seqtk multiqc</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Activate the environment</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate qc-tools</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
</section>
</section>
<section id="directory-structure-setup" class="level2" data-number="2.2">
<h2 data-number="2.2" class="anchored" data-anchor-id="directory-structure-setup"><span class="header-section-number">2.2</span> Directory Structure Setup</h2>
<p>Letâ€™s create an organized directory structure for our analysis:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Navigate to working directory and create structure</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> 02-primary/<span class="dt">{trimmed</span><span class="op">,</span><span class="dt">merged</span><span class="op">,</span><span class="dt">reports</span><span class="op">,</span><span class="dt">kraken2}</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> 02-primary/<span class="dt">{fastQC</span><span class="op">,</span><span class="dt">nanoplot}</span>/<span class="dt">{raw</span><span class="op">,</span><span class="dt">processed}</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> 02-primary/trimmed/sub</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><strong>Directory structure:</strong></p>
<ul>
<li><code>raw/</code>: Raw sequencing data and initial QC reports</li>
<li><code>trimmed/</code>: Filtered and trimmed data with QC reports</li>
<li><code>trimmed/sub/</code>: Subsampled datasets for analysis</li>
<li><code>merged/</code>: Merged Illumina reads (when applicable) with QC</li>
<li><code>kraken2/</code>: Kraken2 database and taxonomic classification results</li>
<li><code>reports/</code>: Comprehensive MultiQC reports</li>
</ul>
</section>
<section id="initial-quality-control-assessment" class="level2" data-number="2.3">
<h2 data-number="2.3" class="anchored" data-anchor-id="initial-quality-control-assessment"><span class="header-section-number">2.3</span> Initial Quality Control Assessment</h2>
<p>Before any preprocessing, we need to understand the characteristics of our raw data. This baseline assessment helps us:</p>
<ul>
<li>Identify potential issues (adapter contamination, quality drops, etc.)</li>
<li>Set appropriate filtering parameters</li>
<li>Evaluate the success of preprocessing steps</li>
</ul>
<section id="illumina-data-qc-with-fastqc" class="level3" data-number="2.3.1">
<h3 data-number="2.3.1" class="anchored" data-anchor-id="illumina-data-qc-with-fastqc"><span class="header-section-number">2.3.1</span> Illumina Data QC with FastQC</h3>
<p>FastQC provides comprehensive quality metrics for short-read data, including per-base quality scores, GC content, adapter content, and sequence duplication levels.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Run FastQC on raw Illumina data</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="ex">fastqc</span> <span class="at">--threads</span> <span class="va">$NCPUS</span> <span class="at">--memory</span> 8000 <span class="dt">\</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">--extract</span> <span class="at">--outdir</span> 02-primary/fastQC/raw/ <span class="dt">\</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>  01-rawdata/<span class="va">${SN}</span>-illumina_R1.fastq.gz 01-rawdata/<span class="va">${SN}</span>-illumina_R2.fastq.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><strong>Key FastQC metrics to examine:</strong></p>
<ul>
<li><strong>Per base sequence quality</strong>: Should be &gt;20 (preferably &gt;30) across most of the read</li>
<li><strong>Per sequence quality scores</strong>: Distribution should be skewed toward high quality</li>
<li><strong>Adapter content</strong>: Should be minimal in good data</li>
<li><strong>Sequence duplication levels</strong>: High duplication may indicate PCR bias</li>
</ul>
</section>
<section id="nanopore-data-qc-with-nanoplot" class="level3" data-number="2.3.2">
<h3 data-number="2.3.2" class="anchored" data-anchor-id="nanopore-data-qc-with-nanoplot"><span class="header-section-number">2.3.2</span> Nanopore Data QC with NanoPlot</h3>
<p>NanoPlot is specifically designed for long-read technologies and provides metrics relevant to Nanopore data characteristics.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Run NanoPlot on raw Nanopore data</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> 02-primary/nanoplot/</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="ex">NanoPlot</span> <span class="at">--fastq</span> 01-rawdata/<span class="va">${SN}</span>-nanopore.fastq.gz <span class="dt">\</span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">--outdir</span> 02-primary/nanoplot/raw/ <span class="at">--info_in_report</span> <span class="dt">\</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">--loglength</span> <span class="at">--threads</span> <span class="va">$NCPUS</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><strong>Key NanoPlot metrics to examine:</strong></p>
<ul>
<li><strong>Read length distribution</strong>: Nanopore typically produces reads from 1kb to &gt;100kb</li>
<li><strong>Quality score distribution</strong>: Lower than Illumina but should show reasonable distribution</li>
<li><strong>Yield over time</strong>: Helps assess sequencing run performance</li>
</ul>
</section>
</section>
<section id="preprocessing-filtering-and-trimming" class="level2" data-number="2.4">
<h2 data-number="2.4" class="anchored" data-anchor-id="preprocessing-filtering-and-trimming"><span class="header-section-number">2.4</span> Preprocessing: Filtering and Trimming</h2>
<p>Based on the initial QC results, weâ€™ll now filter and trim our data to improve quality while retaining maximum useful information.</p>
<section id="illumina-data-preprocessing-with-fastp" class="level3" data-number="2.4.1">
<h3 data-number="2.4.1" class="anchored" data-anchor-id="illumina-data-preprocessing-with-fastp"><span class="header-section-number">2.4.1</span> Illumina Data Preprocessing with fastp</h3>
<p>fastp is an ultra-fast tool that combines multiple preprocessing steps: adapter trimming, quality filtering, and read merging. We provide two filtering options with different stringency levels:</p>
<div class="tabset-margin-container"></div><div class="panel-tabset">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-1-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-1" role="tab" aria-controls="tabset-1-1" aria-selected="true">Option A: Moderate Filtering</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-2" role="tab" aria-controls="tabset-1-2" aria-selected="false">Option B: Stringent Filtering</a></li></ul>
<div class="tab-content">
<div id="tabset-1-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-1-1-tab">
<p><strong>Parameters:</strong> Average quality â‰¥28, minimum length 101bp</p>
<p>This option provides a balance between data retention and quality improvement, suitable for most genome assembly projects.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Run fastp for Illumina paired-end data (Option A)</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="bu">export</span> <span class="va">opt</span><span class="op">=</span>A</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="ex">fastp</span> <span class="dt">\</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">-i</span> 01-rawdata/<span class="va">${SN}</span>-illumina_R1.fastq.gz <span class="at">-I</span> 01-rawdata/<span class="va">${SN}</span>-illumina_R2.fastq.gz <span class="dt">\</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">-o</span> 02-primary/trimmed/<span class="va">${SN}</span>-<span class="va">${opt}</span>-illumina_R1.fastq.gz <span class="dt">\</span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">-O</span> 02-primary/trimmed/<span class="va">${SN}</span>-<span class="va">${opt}</span>-illumina_R2.fastq.gz <span class="dt">\</span></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">--detect_adapter_for_pe</span> <span class="at">--trim_poly_g</span> <span class="at">--trim_poly_x</span> <span class="dt">\</span></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">--average_qual</span> 28 <span class="at">--length_required</span> 101 <span class="dt">\</span></span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">--html</span> 02-primary/trimmed/<span class="va">${SN}</span>-<span class="va">${opt}</span>-illumina_fastp.html <span class="dt">\</span></span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">--cut_right_window_size</span> 4 <span class="at">--cut_right_mean_quality</span> 20 <span class="at">--cut_right</span> <span class="dt">\</span></span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">--cut_tail_window_size</span> 4 <span class="at">--cut_tail_mean_quality</span> 25 <span class="at">--cut_tail</span> <span class="dt">\</span></span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">--thread</span> <span class="va">$NCPUS</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="tabset-1-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-2-tab">
<p><strong>Parameters:</strong> Average quality â‰¥30, minimum length 125bp</p>
<p>This option provides higher quality thresholds for projects requiring maximum data quality, though with potentially higher data loss.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Run fastp for Illumina paired-end data (Option B)</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="bu">export</span> <span class="va">opt</span><span class="op">=</span>B</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="ex">fastp</span> <span class="dt">\</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">-i</span> 01-rawdata/<span class="va">${SN}</span>-illumina_R1.fastq.gz <span class="at">-I</span> 01-rawdata/<span class="va">${SN}</span>-illumina_R2.fastq.gz <span class="dt">\</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">-o</span> 02-primary/trimmed/<span class="va">${SN}</span>-<span class="va">${opt}</span>-illumina_R1.fastq.gz <span class="dt">\</span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">-O</span> 02-primary/trimmed/<span class="va">${SN}</span>-<span class="va">${opt}</span>-illumina_R2.fastq.gz <span class="dt">\</span></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">--detect_adapter_for_pe</span> <span class="at">--trim_poly_g</span> <span class="at">--trim_poly_x</span> <span class="dt">\</span></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">--average_qual</span> 30 <span class="at">--length_required</span> 125 <span class="dt">\</span></span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">--html</span> 02-primary/trimmed/<span class="va">${SN}</span>-<span class="va">${opt}</span>-illumina_fastp.html <span class="dt">\</span></span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">--cut_right_window_size</span> 4 <span class="at">--cut_right_mean_quality</span> 25 <span class="at">--cut_right</span> <span class="dt">\</span></span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">--cut_tail_window_size</span> 4 <span class="at">--cut_tail_mean_quality</span> 28 <span class="at">--cut_tail</span> <span class="dt">\</span></span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">--thread</span> <span class="va">$NCPUS</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<p><strong>fastp parameters explained:</strong></p>
<ul>
<li><code>--detect_adapter_for_pe</code>: Automatically detect and remove adapters</li>
<li><code>--trim_poly_g/x</code>: Remove poly-G/X tails (common in NextSeq/NovaSeq)</li>
<li><code>--average_qual</code>: Minimum average quality score threshold</li>
<li><code>--length_required</code>: Minimum read length after trimming</li>
<li><code>--cut_right/tail</code>: Quality-based trimming from 3â€™ end</li>
</ul>
<p><strong>Choosing between options:</strong> - <strong>Option A</strong> is recommended for most projects, providing good quality while preserving data - <strong>Option B</strong> is suitable when maximum quality is required and sufficient coverage depth is available</p>
</section>
<section id="nanopore-data-preprocessing" class="level3" data-number="2.4.2">
<h3 data-number="2.4.2" class="anchored" data-anchor-id="nanopore-data-preprocessing"><span class="header-section-number">2.4.2</span> Nanopore Data Preprocessing</h3>
<p>For Nanopore data, we focus on length and quality filtering rather than adapter trimming:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Run fastp for Nanopore data (quality and length filtering)</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="ex">fastp</span> <span class="dt">\</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">-i</span> 01-rawdata/<span class="va">${SN}</span>-nanopore.fastq.gz <span class="at">-o</span> 02-primary/trimmed/<span class="va">${SN}</span>-nanopore.fastq.gz <span class="dt">\</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">--disable_adapter_trimming</span> <span class="dt">\</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">--average_qual</span> 16 <span class="at">--length_required</span> 1000 <span class="dt">\</span></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">--html</span> 02-primary/trimmed/<span class="va">${SN}</span>-nanopore_fastp.html <span class="dt">\</span></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">--thread</span> <span class="va">$NCPUS</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><strong>Nanopore-specific parameters:</strong></p>
<ul>
<li><code>--disable_adapter_trimming</code>: Adapters typically removed during basecalling</li>
<li><code>--average_qual 16</code>: More lenient quality threshold for long reads</li>
<li><code>--length_required 1000</code>: Remove very short reads that are less useful for assembly</li>
</ul>
<blockquote class="blockquote">
<p>Note that basecalling and important preliminary steps to clean nanopore reads are not seen here.</p>
</blockquote>
</section>
</section>
<section id="read-merging-illumina-only" class="level2" data-number="2.5">
<h2 data-number="2.5" class="anchored" data-anchor-id="read-merging-illumina-only"><span class="header-section-number">2.5</span> Read Merging (Illumina Only)</h2>
<p>For Illumina paired-end data, overlapping read pairs can be merged to create longer, higher-quality reads. This is particularly beneficial for genome assembly.</p>
<div class="tabset-margin-container"></div><div class="panel-tabset">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-2-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-2-1" role="tab" aria-controls="tabset-2-1" aria-selected="true">Option A: Moderate Filtering</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-2-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-2-2" role="tab" aria-controls="tabset-2-2" aria-selected="false">Option B: Stringent Filtering</a></li></ul>
<div class="tab-content">
<div id="tabset-2-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-2-1-tab">
<div class="sourceCode" id="cb11"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Merge overlapping Illumina read pairs (Option A)</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="bu">export</span> <span class="va">opt</span><span class="op">=</span>A</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="ex">fastp</span> <span class="dt">\</span></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">--merge</span> <span class="at">--correction</span> <span class="at">--merged_out</span> 02-primary/merged/<span class="va">${SN}</span>-<span class="va">${opt}</span>-illumina.fastq.gz <span class="dt">\</span></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">--disable_quality_filtering</span> <span class="dt">\</span></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">-i</span> 02-primary/trimmed/<span class="va">${SN}</span>-<span class="va">${opt}</span>-illumina_R1.fastq.gz <span class="dt">\</span></span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">-I</span> 02-primary/trimmed/<span class="va">${SN}</span>-<span class="va">${opt}</span>-illumina_R2.fastq.gz <span class="dt">\</span></span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">-o</span> 02-primary/merged/<span class="va">${SN}</span>-<span class="va">${opt}</span>-illumina_R1.fastq.gz <span class="dt">\</span></span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">-O</span> 02-primary/merged/<span class="va">${SN}</span>-<span class="va">${opt}</span>-illumina_R2.fastq.gz <span class="dt">\</span></span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">--detect_adapter_for_pe</span> <span class="at">--trim_poly_g</span> <span class="at">--trim_poly_x</span> <span class="dt">\</span></span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">--html</span> merged/<span class="va">${SN}</span>-<span class="va">${opt}</span>-illumina_merge_fastp.html <span class="dt">\</span></span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">--thread</span> <span class="va">$NCPUS</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="tabset-2-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-2-2-tab">
<div class="sourceCode" id="cb12"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Merge overlapping Illumina read pairs (Option B)</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="bu">export</span> <span class="va">opt</span><span class="op">=</span>B</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a><span class="ex">fastp</span> <span class="dt">\</span></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">--merge</span> <span class="at">--correction</span> <span class="at">--merged_out</span> 02-primary/merged/<span class="va">${SN}</span>-<span class="va">${opt}</span>-illumina.fastq.gz <span class="dt">\</span></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">--disable_quality_filtering</span> <span class="dt">\</span></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">-i</span> 02-primary/trimmed/<span class="va">${SN}</span>-<span class="va">${opt}</span>-illumina_R1.fastq.gz <span class="dt">\</span></span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">-I</span> 02-primary/trimmed/<span class="va">${SN}</span>-<span class="va">${opt}</span>-illumina_R2.fastq.gz <span class="dt">\</span></span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">-o</span> 02-primary/merged/<span class="va">${SN}</span>-<span class="va">${opt}</span>-illumina_R1.fastq.gz <span class="dt">\</span></span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">-O</span> 02-primary/merged/<span class="va">${SN}</span>-<span class="va">${opt}</span>-illumina_R2.fastq.gz <span class="dt">\</span></span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">--detect_adapter_for_pe</span> <span class="at">--trim_poly_g</span> <span class="at">--trim_poly_x</span> <span class="dt">\</span></span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">--html</span> merged/<span class="va">${SN}</span>-<span class="va">${opt}</span>-illumina_merge_fastp.html <span class="dt">\</span></span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">--thread</span> <span class="va">$NCPUS</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<p><strong>Merging benefits:</strong></p>
<ul>
<li>Creates longer reads from overlapping pairs</li>
<li>Improves base quality through consensus</li>
<li>Reduces data complexity for assembly algorithms</li>
<li>Maintains unmerged pairs for additional coverage</li>
</ul>
</section>
<section id="post-processing-quality-control" class="level2" data-number="2.6">
<h2 data-number="2.6" class="anchored" data-anchor-id="post-processing-quality-control"><span class="header-section-number">2.6</span> Post-Processing Quality Control</h2>
<p>After preprocessing, we need to verify that our filtering and trimming steps improved data quality without excessive data loss.</p>
<section id="post-processing-qc-for-illumina-data" class="level3" data-number="2.6.1">
<h3 data-number="2.6.1" class="anchored" data-anchor-id="post-processing-qc-for-illumina-data"><span class="header-section-number">2.6.1</span> Post-Processing QC for Illumina Data</h3>
<div class="sourceCode" id="cb13"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="ex">fastqc</span> <span class="at">--threads</span> <span class="va">$NCPUS</span> <span class="at">--memory</span> 8000 <span class="dt">\</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">--extract</span> <span class="at">--outdir</span> 02-primary/fastQC/processed <span class="dt">\</span></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>  02-primary/merged/<span class="pp">*</span>illumina<span class="pp">*</span>.fastq.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="post-processing-qc-for-nanopore-data" class="level3" data-number="2.6.2">
<h3 data-number="2.6.2" class="anchored" data-anchor-id="post-processing-qc-for-nanopore-data"><span class="header-section-number">2.6.2</span> Post-Processing QC for Nanopore Data</h3>
<div class="sourceCode" id="cb14"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="ex">NanoPlot</span> <span class="at">--fastq</span> 02-primary/trimmed/<span class="va">${SN}</span>-nanopore.fastq.gz <span class="dt">\</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">--outdir</span> 02-primary/nanoplot/processed <span class="at">--info_in_report</span> <span class="dt">\</span></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">--loglength</span> <span class="at">--threads</span> <span class="va">$NCPUS</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
</section>
<section id="taxonomic-classification-with-kraken2" class="level2" data-number="2.7">
<h2 data-number="2.7" class="anchored" data-anchor-id="taxonomic-classification-with-kraken2"><span class="header-section-number">2.7</span> Taxonomic Classification with Kraken2</h2>
<p>Taxonomic classification helps identify potential contamination and provides insights into the composition of your sequencing data. Weâ€™ll use Kraken2, a fast and accurate taxonomic classifier, on both full datasets and subsampled data to demonstrate scalable analysis approaches.</p>
<section id="database-setup" class="level3" data-number="2.7.1">
<h3 data-number="2.7.1" class="anchored" data-anchor-id="database-setup"><span class="header-section-number">2.7.1</span> Database Setup</h3>
<p>First, we need to download and set up the Kraken2 database. Weâ€™ll use the PlusPF database which includes bacteria, archaea, viruses, fungi, and protozoa:</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Download and extract Kraken2 database</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> 02-primary/kraken2</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> k2_pluspf_08gb_20250402</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> k2_pluspf_08gb_20250402</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> https://genome-idx.s3.amazonaws.com/kraken/k2_pluspf_08gb_20250402.tar.gz</span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a><span class="fu">tar</span> <span class="at">-xzf</span> k2_pluspf_08gb_20250402.tar.gz</span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span> k2_pluspf_08gb_20250402.tar.gz</span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ../../..</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<blockquote class="blockquote">
<p>Note that we are using a capped database limited to 8Gb (in place of ~70). Therefore sensitivity will not be the best. In real world scenario itâ€™s recommended to use the complete database.</p>
</blockquote>
</section>
<section id="subsampling-with-seqtk" class="level3" data-number="2.7.2">
<h3 data-number="2.7.2" class="anchored" data-anchor-id="subsampling-with-seqtk"><span class="header-section-number">2.7.2</span> Subsampling with seqtk</h3>
<p>For demonstration purposes and faster analysis, weâ€™ll create subsampled datasets with 100,000 reads each. This approach is useful for:</p>
<ul>
<li>Quick preliminary analysis</li>
<li>Testing workflows before full-scale analysis</li>
<li>Reducing computational requirements for large datasets</li>
</ul>
<div class="sourceCode" id="cb16"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create subsamples of trimmed Illumina data (100k reads each)</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Set seed for reproducible subsampling</span></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a><span class="bu">export</span> <span class="va">SEED</span><span class="op">=</span>42</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a><span class="bu">export</span> <span class="va">opt</span><span class="op">=</span>A</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Subsample Option A datasets</span></span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a><span class="ex">seqtk</span> sample <span class="at">-s</span> <span class="va">$SEED</span> 02-primary/trimmed/<span class="va">${SN}</span>-<span class="va">${opt}</span>-illumina_R1.fastq.gz 100000 <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gzip</span> <span class="op">&gt;</span> 02-primary/trimmed/sub/<span class="va">${SN}</span>-<span class="va">${opt}</span>-illumina_R1_sub100k.fastq.gz</span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a><span class="ex">seqtk</span> sample <span class="at">-s</span> <span class="va">$SEED</span> 02-primary/trimmed/<span class="va">${SN}</span>-<span class="va">${opt}</span>-illumina_R2.fastq.gz 100000 <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gzip</span> <span class="op">&gt;</span> 02-primary/trimmed/sub/<span class="va">${SN}</span>-<span class="va">${opt}</span>-illumina_R2_sub100k.fastq.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<blockquote class="blockquote">
<p>Note that only 1 dataset is enough.</p>
</blockquote>
<p><strong>seqtk parameters explained:</strong></p>
<ul>
<li><code>sample -s SEED</code>: Random sampling with fixed seed for reproducibility</li>
<li><code>100000</code>: Number of reads to sample</li>
<li>Output is compressed with gzip for space efficiency</li>
</ul>
</section>
<section id="kraken2-classification" class="level3" data-number="2.7.3">
<h3 data-number="2.7.3" class="anchored" data-anchor-id="kraken2-classification"><span class="header-section-number">2.7.3</span> Kraken2 Classification</h3>
<p>Now weâ€™ll run Kraken2 on subsampled dataset to demonstrate the workflow scalability:</p>
<section id="subsampled-datasets" class="level4" data-number="2.7.3.1">
<h4 data-number="2.7.3.1" class="anchored" data-anchor-id="subsampled-datasets"><span class="header-section-number">2.7.3.1</span> Subsampled Datasets</h4>
<div class="sourceCode" id="cb17"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Run Kraken2 on subsampled datasets (faster analysis)</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="bu">export</span> <span class="va">opt</span><span class="op">=</span>A</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a><span class="ex">kraken2</span> <span class="at">--db</span> 02-primary/kraken2/k2_pluspf_08gb_20250402 <span class="dt">\</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">--threads</span> <span class="va">$NCPUS</span> <span class="at">--paired</span> <span class="dt">\</span></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">--output</span> 02-primary/kraken2/<span class="va">${SN}</span>-<span class="va">${opt}</span>-illumina_sub100k.kraken <span class="dt">\</span></span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">--report</span> 02-primary/kraken2/<span class="va">${SN}</span>-<span class="va">${opt}</span>-illumina_sub100k.report <span class="dt">\</span></span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>  02-primary/trimmed/sub/<span class="va">${SN}</span>-<span class="va">${opt}</span>-illumina_R1_sub100k.fastq.gz <span class="dt">\</span></span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>  02-primary/trimmed/sub/<span class="va">${SN}</span>-<span class="va">${opt}</span>-illumina_R2_sub100k.fastq.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><strong>Kraken2 parameters explained:</strong></p>
<ul>
<li><code>--db</code>: Path to Kraken2 database</li>
<li><code>--threads</code>: Number of CPU threads to use</li>
<li><code>--paired</code>: Indicates paired-end reads</li>
<li><code>--output</code>: Detailed classification output (optional, can be large)</li>
<li><code>--report</code>: Summary report with taxonomic composition</li>
</ul>
</section>
</section>
<section id="interpreting-kraken2-results" class="level3" data-number="2.7.4">
<h3 data-number="2.7.4" class="anchored" data-anchor-id="interpreting-kraken2-results"><span class="header-section-number">2.7.4</span> Interpreting Kraken2 Results</h3>
<p>The Kraken2 report provides taxonomic composition with the following columns:</p>
<ol type="1">
<li><strong>Percentage of reads</strong>: Fraction of reads assigned to this taxon</li>
<li><strong>Number of reads</strong>: Absolute count of reads assigned</li>
<li><strong>Number of reads at this level</strong>: Reads assigned specifically to this taxon</li>
<li><strong>Taxonomic rank</strong>: D (domain), K (kingdom), P (phylum), C (class), O (order), F (family), G (genus), S (species)</li>
<li><strong>NCBI taxonomy ID</strong>: Unique identifier for the taxon</li>
<li><strong>Scientific name</strong>: Taxonomic name</li>
</ol>
<p><strong>Key metrics to examine:</strong></p>
<ul>
<li><strong>Unclassified reads</strong>: High percentages may indicate novel organisms or poor database coverage</li>
<li><strong>Dominant taxa</strong>: Should align with expected organism(s)</li>
<li><strong>Contamination indicators</strong>: Unexpected taxa may suggest contamination</li>
</ul>
</section>
</section>
<section id="comprehensive-reporting-with-multiqc" class="level2" data-number="2.8">
<h2 data-number="2.8" class="anchored" data-anchor-id="comprehensive-reporting-with-multiqc"><span class="header-section-number">2.8</span> Comprehensive Reporting with MultiQC</h2>
<p>MultiQC aggregates results from multiple QC tools into a single, interactive HTML report, making it easy to compare before/after preprocessing results.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Generate comprehensive MultiQC report</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="ex">multiqc</span> 02-primary/ <span class="at">--outdir</span> 02-primary/reports/ <span class="at">--dirs</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><strong>MultiQC benefits:</strong></p>
<ul>
<li>Combines FastQC, NanoPlot, kraken2, and fastp reports</li>
<li>Provides side-by-side comparisons</li>
<li>Interactive plots for detailed exploration</li>
<li>Summary statistics across all samples</li>
</ul>
</section>
<section id="expected-outcomes" class="level2" data-number="2.9">
<h2 data-number="2.9" class="anchored" data-anchor-id="expected-outcomes"><span class="header-section-number">2.9</span> Expected Outcomes</h2>
<p>After completing this primary analysis workflow, you should observe:</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="ex">02-primary</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="ex">â”œâ”€â”€</span> fastQC</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a><span class="ex">â”‚</span>   â”œâ”€â”€ processed</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a><span class="ex">â”‚</span>   â”‚   â”œâ”€â”€ tuto482-A-illumina_fastqc/</span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a><span class="ex">â”‚</span>   â”‚   â”œâ”€â”€ tuto482-A-illumina_fastqc.html</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a><span class="ex">â”‚</span>   â”‚   â”œâ”€â”€ tuto482-A-illumina_fastqc.zip</span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a><span class="ex">â”‚</span>   â”‚   â”œâ”€â”€ tuto482-A-illumina_R1_fastqc/</span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a><span class="ex">â”‚</span>   â”‚   â”œâ”€â”€ tuto482-A-illumina_R1_fastqc.html</span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a><span class="ex">â”‚</span>   â”‚   â”œâ”€â”€ tuto482-A-illumina_R1_fastqc.zip</span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a><span class="ex">â”‚</span>   â”‚   â”œâ”€â”€ tuto482-A-illumina_R2_fastqc/</span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a><span class="ex">â”‚</span>   â”‚   â”œâ”€â”€ tuto482-A-illumina_R2_fastqc.html</span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a><span class="ex">â”‚</span>   â”‚   â”œâ”€â”€ tuto482-A-illumina_R2_fastqc.zip</span>
<span id="cb19-13"><a href="#cb19-13" aria-hidden="true" tabindex="-1"></a><span class="ex">â”‚</span>   â”‚   â”œâ”€â”€ tuto482-B-illumina_fastqc/</span>
<span id="cb19-14"><a href="#cb19-14" aria-hidden="true" tabindex="-1"></a><span class="ex">â”‚</span>   â”‚   â”œâ”€â”€ tuto482-B-illumina_fastqc.html</span>
<span id="cb19-15"><a href="#cb19-15" aria-hidden="true" tabindex="-1"></a><span class="ex">â”‚</span>   â”‚   â”œâ”€â”€ tuto482-B-illumina_fastqc.zip</span>
<span id="cb19-16"><a href="#cb19-16" aria-hidden="true" tabindex="-1"></a><span class="ex">â”‚</span>   â”‚   â”œâ”€â”€ tuto482-B-illumina_R1_fastqc/</span>
<span id="cb19-17"><a href="#cb19-17" aria-hidden="true" tabindex="-1"></a><span class="ex">â”‚</span>   â”‚   â”œâ”€â”€ tuto482-B-illumina_R1_fastqc.html</span>
<span id="cb19-18"><a href="#cb19-18" aria-hidden="true" tabindex="-1"></a><span class="ex">â”‚</span>   â”‚   â”œâ”€â”€ tuto482-B-illumina_R1_fastqc.zip</span>
<span id="cb19-19"><a href="#cb19-19" aria-hidden="true" tabindex="-1"></a><span class="ex">â”‚</span>   â”‚   â”œâ”€â”€ tuto482-B-illumina_R2_fastqc/</span>
<span id="cb19-20"><a href="#cb19-20" aria-hidden="true" tabindex="-1"></a><span class="ex">â”‚</span>   â”‚   â”œâ”€â”€ tuto482-B-illumina_R2_fastqc.html</span>
<span id="cb19-21"><a href="#cb19-21" aria-hidden="true" tabindex="-1"></a><span class="ex">â”‚</span>   â”‚   â””â”€â”€ tuto482-B-illumina_R2_fastqc.zip</span>
<span id="cb19-22"><a href="#cb19-22" aria-hidden="true" tabindex="-1"></a><span class="ex">â”‚</span>   â””â”€â”€ raw</span>
<span id="cb19-23"><a href="#cb19-23" aria-hidden="true" tabindex="-1"></a><span class="ex">â”‚</span>       â”œâ”€â”€ tuto482-illumina_R1_fastqc</span>
<span id="cb19-24"><a href="#cb19-24" aria-hidden="true" tabindex="-1"></a><span class="ex">â”‚</span>       â”œâ”€â”€ tuto482-illumina_R1_fastqc.html</span>
<span id="cb19-25"><a href="#cb19-25" aria-hidden="true" tabindex="-1"></a><span class="ex">â”‚</span>       â”œâ”€â”€ tuto482-illumina_R1_fastqc.zip</span>
<span id="cb19-26"><a href="#cb19-26" aria-hidden="true" tabindex="-1"></a><span class="ex">â”‚</span>       â”œâ”€â”€ tuto482-illumina_R2_fastqc/</span>
<span id="cb19-27"><a href="#cb19-27" aria-hidden="true" tabindex="-1"></a><span class="ex">â”‚</span>       â”œâ”€â”€ tuto482-illumina_R2_fastqc.html</span>
<span id="cb19-28"><a href="#cb19-28" aria-hidden="true" tabindex="-1"></a><span class="ex">â”‚</span>       â””â”€â”€ tuto482-illumina_R2_fastqc.zip</span>
<span id="cb19-29"><a href="#cb19-29" aria-hidden="true" tabindex="-1"></a><span class="ex">â”œâ”€â”€</span> merged</span>
<span id="cb19-30"><a href="#cb19-30" aria-hidden="true" tabindex="-1"></a><span class="ex">â”‚</span>   â”œâ”€â”€ tuto482-A-illumina.fastq.gz</span>
<span id="cb19-31"><a href="#cb19-31" aria-hidden="true" tabindex="-1"></a><span class="ex">â”‚</span>   â”œâ”€â”€ tuto482-A-illumina_R1.fastq.gz</span>
<span id="cb19-32"><a href="#cb19-32" aria-hidden="true" tabindex="-1"></a><span class="ex">â”‚</span>   â”œâ”€â”€ tuto482-A-illumina_R2.fastq.gz</span>
<span id="cb19-33"><a href="#cb19-33" aria-hidden="true" tabindex="-1"></a><span class="ex">â”‚</span>   â”œâ”€â”€ tuto482-B-illumina.fastq.gz</span>
<span id="cb19-34"><a href="#cb19-34" aria-hidden="true" tabindex="-1"></a><span class="ex">â”‚</span>   â”œâ”€â”€ tuto482-B-illumina_R1.fastq.gz</span>
<span id="cb19-35"><a href="#cb19-35" aria-hidden="true" tabindex="-1"></a><span class="ex">â”‚</span>   â””â”€â”€ tuto482-B-illumina_R2.fastq.gz</span>
<span id="cb19-36"><a href="#cb19-36" aria-hidden="true" tabindex="-1"></a><span class="ex">â”œâ”€â”€</span> nanoplot</span>
<span id="cb19-37"><a href="#cb19-37" aria-hidden="true" tabindex="-1"></a><span class="ex">â”‚</span>   â”œâ”€â”€ processed</span>
<span id="cb19-38"><a href="#cb19-38" aria-hidden="true" tabindex="-1"></a><span class="ex">â”‚</span>   â”‚   â”œâ”€â”€ ...</span>
<span id="cb19-39"><a href="#cb19-39" aria-hidden="true" tabindex="-1"></a><span class="ex">â”‚</span>   â”‚   â”œâ”€â”€ NanoPlot-report.html</span>
<span id="cb19-40"><a href="#cb19-40" aria-hidden="true" tabindex="-1"></a><span class="ex">â”‚</span>   â”‚   â”œâ”€â”€ NanoStats.txt</span>
<span id="cb19-41"><a href="#cb19-41" aria-hidden="true" tabindex="-1"></a><span class="ex">â”‚</span>   â”‚   â”œâ”€â”€ ...</span>
<span id="cb19-42"><a href="#cb19-42" aria-hidden="true" tabindex="-1"></a><span class="ex">â”‚</span>   â””â”€â”€ raw</span>
<span id="cb19-43"><a href="#cb19-43" aria-hidden="true" tabindex="-1"></a><span class="ex">â”‚</span>       â”œâ”€â”€ ...</span>
<span id="cb19-44"><a href="#cb19-44" aria-hidden="true" tabindex="-1"></a><span class="ex">â”‚</span>       â”œâ”€â”€ NanoPlot-report.html</span>
<span id="cb19-45"><a href="#cb19-45" aria-hidden="true" tabindex="-1"></a><span class="ex">â”‚</span>       â”œâ”€â”€ NanoStats.txt</span>
<span id="cb19-46"><a href="#cb19-46" aria-hidden="true" tabindex="-1"></a><span class="ex">â”‚</span>       â”œâ”€â”€ ...</span>
<span id="cb19-47"><a href="#cb19-47" aria-hidden="true" tabindex="-1"></a><span class="ex">â”œâ”€â”€</span> reports</span>
<span id="cb19-48"><a href="#cb19-48" aria-hidden="true" tabindex="-1"></a><span class="ex">â”‚</span>   â”œâ”€â”€ multiqc_data/</span>
<span id="cb19-49"><a href="#cb19-49" aria-hidden="true" tabindex="-1"></a><span class="ex">â”‚</span>   â””â”€â”€ multiqc_report.html</span>
<span id="cb19-50"><a href="#cb19-50" aria-hidden="true" tabindex="-1"></a><span class="ex">â””â”€â”€</span> trimmed</span>
<span id="cb19-51"><a href="#cb19-51" aria-hidden="true" tabindex="-1"></a>    <span class="ex">â”œâ”€â”€</span> sub/</span>
<span id="cb19-52"><a href="#cb19-52" aria-hidden="true" tabindex="-1"></a>    <span class="ex">â”œâ”€â”€</span> tuto482-A-illumina_fastp.html</span>
<span id="cb19-53"><a href="#cb19-53" aria-hidden="true" tabindex="-1"></a>    <span class="ex">â”œâ”€â”€</span> tuto482-A-illumina_R1.fastq.gz</span>
<span id="cb19-54"><a href="#cb19-54" aria-hidden="true" tabindex="-1"></a>    <span class="ex">â”œâ”€â”€</span> tuto482-A-illumina_R2.fastq.gz</span>
<span id="cb19-55"><a href="#cb19-55" aria-hidden="true" tabindex="-1"></a>    <span class="ex">â”œâ”€â”€</span> tuto482-B-illumina_fastp.html</span>
<span id="cb19-56"><a href="#cb19-56" aria-hidden="true" tabindex="-1"></a>    <span class="ex">â”œâ”€â”€</span> tuto482-B-illumina_R1.fastq.gz</span>
<span id="cb19-57"><a href="#cb19-57" aria-hidden="true" tabindex="-1"></a>    <span class="ex">â”œâ”€â”€</span> tuto482-B-illumina_R2.fastq.gz</span>
<span id="cb19-58"><a href="#cb19-58" aria-hidden="true" tabindex="-1"></a>    <span class="ex">â”œâ”€â”€</span> tuto482-nanopore_fastp.html</span>
<span id="cb19-59"><a href="#cb19-59" aria-hidden="true" tabindex="-1"></a>    <span class="ex">â””â”€â”€</span> tuto482-nanopore.fastq.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<section id="illumina-data-improvements" class="level3" data-number="2.9.1">
<h3 data-number="2.9.1" class="anchored" data-anchor-id="illumina-data-improvements"><span class="header-section-number">2.9.1</span> Illumina Data Improvements:</h3>
<ul>
<li><strong>Increased average quality scores</strong> across read positions</li>
<li><strong>Reduced adapter contamination</strong> to near-zero levels</li>
</ul>
</section>
<section id="nanopore-data-improvements" class="level3" data-number="2.9.2">
<h3 data-number="2.9.2" class="anchored" data-anchor-id="nanopore-data-improvements"><span class="header-section-number">2.9.2</span> Nanopore Data Improvements:</h3>
<ul>
<li><strong>Removal of very short reads</strong> that donâ€™t contribute to assembly</li>
<li><strong>Improved average quality</strong> while maintaining read length advantages</li>
</ul>
</section>
</section>
<section id="next-steps" class="level2" data-number="2.10">
<h2 data-number="2.10" class="anchored" data-anchor-id="next-steps"><span class="header-section-number">2.10</span> Next Steps</h2>
<p>With high-quality, preprocessed sequencing data, weâ€™re now ready to proceed to:</p>
<ol type="1">
<li><strong>K-mer analysis</strong> to estimate genome characteristics</li>
<li><strong>De novo genome assembly</strong> using hybrid approaches</li>
<li><strong>Assembly quality assessment</strong> and validation</li>
</ol>
<p>The cleaned datasets will serve as input for downstream assembly algorithms, where the improved data quality will directly translate to better assembly contiguity and accuracy.</p>
</section>
<section id="multiple-datasets-for-assembly" class="level2" data-number="2.11">
<h2 data-number="2.11" class="anchored" data-anchor-id="multiple-datasets-for-assembly"><span class="header-section-number">2.11</span> Multiple datasets for assembly</h2>
<p>In de novo genome assembly, itâ€™s considered best practice to generate multiple processed datasets using different preprocessing parameters as input for assembly. This approach leverages the fundamental trade-off between data quality and sequencing depth.</p>
<section id="the-quality-vs.-depth-trade-off" class="level3" data-number="2.11.1">
<h3 data-number="2.11.1" class="anchored" data-anchor-id="the-quality-vs.-depth-trade-off"><span class="header-section-number">2.11.1</span> The Quality vs.&nbsp;Depth Trade-off</h3>
<p>Different preprocessing stringency levels create datasets with distinct characteristics:</p>
<ul>
<li><strong>Less stringent filtering (Option A)</strong>: Higher sequencing depth but potentially lower average quality</li>
<li><strong>More stringent filtering (Option B)</strong>: Higher average quality but reduced sequencing depth</li>
</ul>
<p>The optimal balance depends on several genome-specific factors.</p>
</section>
<section id="sequencing-depth-analysis" class="level3" data-number="2.11.2">
<h3 data-number="2.11.2" class="anchored" data-anchor-id="sequencing-depth-analysis"><span class="header-section-number">2.11.2</span> Sequencing depth analysis</h3>
<ul>
<li><strong>What is the expected depth for Illumina data (A/B)?</strong>
<ul>
<li>Calculate: (Total bases sequenced) / (Estimated genome size)</li>
<li>Recommended minimum: 30-50x for short reads (100x is ideal); 10x-30x for long read</li>
</ul></li>
</ul>


</section>
</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "î§‹";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("https:\/\/nexomis\.github\.io\/tuto-euk-de-novo-genome\/");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




</body></html>