<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.32">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Julien Fouret">

<title>Scaffolding with Ragout – De Novo Genome Assembly Tutorial</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-37eea08aefeeee20ff55810ff984fec1.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap-3fced2653c1411365395c6308d7f7025.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="../styles.css">
</head>

<body class="floating nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">De Novo Genome Assembly Tutorial</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-primary-analysis" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Primary Analysis</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-primary-analysis">    
        <li>
    <a class="dropdown-item" href="../primary-analysis/base-calling-qc.html">
 <span class="dropdown-text">Base Calling &amp; QC</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../primary-analysis/preliminary-analysis.html">
 <span class="dropdown-text">Preliminary Analysis</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-secondary-analysis" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Secondary Analysis</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-secondary-analysis">    
        <li>
    <a class="dropdown-item" href="../secondary-analysis/concepts-genome-assembly.html">
 <span class="dropdown-text">Genome Assembly Concepts</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../secondary-analysis/contigs-de-novo-assembly.html">
 <span class="dropdown-text">Contigs de novo Assembly</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../secondary-analysis/scaffolding.html">
 <span class="dropdown-text">Scaffolding</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../secondary-analysis/assembly-evaluation.html">
 <span class="dropdown-text">Assembly Evaluation</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-tertiary-analysis" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Tertiary Analysis</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-tertiary-analysis">    
        <li>
    <a class="dropdown-item" href="../tertiary-analysis/genome-annotation.html">
 <span class="dropdown-text">Genome Annotation</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../tertiary-analysis/comparative-genomics.html">
 <span class="dropdown-text">Comparative Genomics</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-more" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">More</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-more">    
        <li>
    <a class="dropdown-item" href="../more/reference-enabled-bioinformatics.html">
 <span class="dropdown-text">Reference-enabled Bioinformatics</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../more/intro-nextflow.html">
 <span class="dropdown-text">Intro to NextFlow</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-hands-on-practice" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Hands-on Practice</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-hands-on-practice">    
        <li>
    <a class="dropdown-item" href="../hands-on/01-fetch-public-datasets.html">
 <span class="dropdown-text">Fetch Public Datasets</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../hands-on/02-primary-analysis.html">
 <span class="dropdown-text">Primary analysis</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../hands-on/03-kmer-analysis.html">
 <span class="dropdown-text">Kmer-based Analysis</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../hands-on/04-hybrid-assembly-spades-masurca.html">
 <span class="dropdown-text">Hybrid Assembly (SPAdes/MaSuRCA)</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../hands-on/05-assembly-evaluation.html">
 <span class="dropdown-text">Assembly Evaluation</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../hands-on/06-scaffolding.html">
 <span class="dropdown-text">Secondary scaffolding</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../hands-on/07-repeats-analysis.html">
 <span class="dropdown-text">Repeats Annotation and Masking</span></a>
  </li>  
    </ul>
  </li>
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../about.html"> 
<span class="menu-text">About</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#scaffolding-with-ragout" id="toc-scaffolding-with-ragout" class="nav-link active" data-scroll-target="#scaffolding-with-ragout"><span class="header-section-number">6</span> Scaffolding with Ragout</a>
  <ul class="collapse">
  <li><a href="#configuration-and-setup" id="toc-configuration-and-setup" class="nav-link" data-scroll-target="#configuration-and-setup"><span class="header-section-number">6.1</span> Configuration and Setup</a>
  <ul class="collapse">
  <li><a href="#environment-variables" id="toc-environment-variables" class="nav-link" data-scroll-target="#environment-variables"><span class="header-section-number">6.1.1</span> Environment Variables</a></li>
  <li><a href="#software-dependencies" id="toc-software-dependencies" class="nav-link" data-scroll-target="#software-dependencies"><span class="header-section-number">6.1.2</span> Software Dependencies</a></li>
  <li><a href="#directory-structure-setup" id="toc-directory-structure-setup" class="nav-link" data-scroll-target="#directory-structure-setup"><span class="header-section-number">6.1.3</span> Directory Structure Setup</a></li>
  </ul></li>
  <li><a href="#data-preparation" id="toc-data-preparation" class="nav-link" data-scroll-target="#data-preparation"><span class="header-section-number">6.2</span> Data Preparation</a>
  <ul class="collapse">
  <li><a href="#download-reference-genomes" id="toc-download-reference-genomes" class="nav-link" data-scroll-target="#download-reference-genomes"><span class="header-section-number">6.2.1</span> Download Reference Genomes</a></li>
  <li><a href="#prepare-ragout-recipe-file" id="toc-prepare-ragout-recipe-file" class="nav-link" data-scroll-target="#prepare-ragout-recipe-file"><span class="header-section-number">6.2.2</span> Prepare Ragout Recipe File</a></li>
  </ul></li>
  <li><a href="#ragout-with-sibelia-default-method" id="toc-ragout-with-sibelia-default-method" class="nav-link" data-scroll-target="#ragout-with-sibelia-default-method"><span class="header-section-number">6.3</span> Ragout with Sibelia (Default Method)</a></li>
  <li><a href="#ragout-with-cactus-hal-alignment-method" id="toc-ragout-with-cactus-hal-alignment-method" class="nav-link" data-scroll-target="#ragout-with-cactus-hal-alignment-method"><span class="header-section-number">6.4</span> Ragout with Cactus (HAL Alignment Method)</a>
  <ul class="collapse">
  <li><a href="#software-dependencies-for-cactus-and-jolytree" id="toc-software-dependencies-for-cactus-and-jolytree" class="nav-link" data-scroll-target="#software-dependencies-for-cactus-and-jolytree"><span class="header-section-number">6.4.1</span> Software Dependencies for Cactus and JolyTree</a></li>
  <li><a href="#generate-phylogenetic-tree-with-jolytree" id="toc-generate-phylogenetic-tree-with-jolytree" class="nav-link" data-scroll-target="#generate-phylogenetic-tree-with-jolytree"><span class="header-section-number">6.4.2</span> Generate Phylogenetic Tree with JolyTree</a></li>
  <li><a href="#manual-tree-editing-for-cactus-compatibility" id="toc-manual-tree-editing-for-cactus-compatibility" class="nav-link" data-scroll-target="#manual-tree-editing-for-cactus-compatibility"><span class="header-section-number">6.4.3</span> Manual Tree Editing for Cactus Compatibility</a></li>
  <li><a href="#create-cactus-sequence-file" id="toc-create-cactus-sequence-file" class="nav-link" data-scroll-target="#create-cactus-sequence-file"><span class="header-section-number">6.4.4</span> Create Cactus Sequence File</a></li>
  <li><a href="#run-cactus-to-generate-hal-alignment" id="toc-run-cactus-to-generate-hal-alignment" class="nav-link" data-scroll-target="#run-cactus-to-generate-hal-alignment"><span class="header-section-number">6.4.5</span> Run Cactus to Generate HAL Alignment</a></li>
  <li><a href="#convert-hal-to-maf-format" id="toc-convert-hal-to-maf-format" class="nav-link" data-scroll-target="#convert-hal-to-maf-format"><span class="header-section-number">6.4.6</span> Convert HAL to MAF Format</a></li>
  <li><a href="#create-ragout-recipe-for-maf-method" id="toc-create-ragout-recipe-for-maf-method" class="nav-link" data-scroll-target="#create-ragout-recipe-for-maf-method"><span class="header-section-number">6.4.7</span> Create Ragout Recipe for MAF Method</a></li>
  <li><a href="#run-ragout-with-maf-alignment" id="toc-run-ragout-with-maf-alignment" class="nav-link" data-scroll-target="#run-ragout-with-maf-alignment"><span class="header-section-number">6.4.8</span> Run Ragout with MAF Alignment</a></li>
  </ul></li>
  <li><a href="#method-comparison" id="toc-method-comparison" class="nav-link" data-scroll-target="#method-comparison"><span class="header-section-number">6.5</span> Method Comparison</a></li>
  <li><a href="#rna-seq-based-scaffolding-with-agouti" id="toc-rna-seq-based-scaffolding-with-agouti" class="nav-link" data-scroll-target="#rna-seq-based-scaffolding-with-agouti"><span class="header-section-number">6.6</span> RNA-Seq Based Scaffolding with Agouti</a></li>
  <li><a href="#expected-outcomes" id="toc-expected-outcomes" class="nav-link" data-scroll-target="#expected-outcomes"><span class="header-section-number">6.7</span> Expected Outcomes</a>
  <ul class="collapse">
  <li><a href="#for-sibelia-method-06-scaffoldingragout_output" id="toc-for-sibelia-method-06-scaffoldingragout_output" class="nav-link" data-scroll-target="#for-sibelia-method-06-scaffoldingragout_output"><span class="header-section-number">6.7.1</span> For Sibelia Method (<code>06-scaffolding/ragout_output/</code>):</a></li>
  <li><a href="#for-cactus-method-06-scaffoldingragout_maf_output" id="toc-for-cactus-method-06-scaffoldingragout_maf_output" class="nav-link" data-scroll-target="#for-cactus-method-06-scaffoldingragout_maf_output"><span class="header-section-number">6.7.2</span> For Cactus Method (<code>06-scaffolding/ragout_maf_output/</code>):</a></li>
  <li><a href="#additional-cactus-outputs-06-scaffoldingcactus" id="toc-additional-cactus-outputs-06-scaffoldingcactus" class="nav-link" data-scroll-target="#additional-cactus-outputs-06-scaffoldingcactus"><span class="header-section-number">6.7.3</span> Additional Cactus Outputs (<code>06-scaffolding/cactus/</code>):</a></li>
  </ul></li>
  </ul></li>
  </ul>
</nav>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar zindex-bottom">
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Scaffolding with Ragout</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Julien Fouret </p>
          </div>
  </div>
    
  
    
  </div>
  


</header>


<section id="scaffolding-with-ragout" class="level1" data-number="6">
<h1 data-number="6"><span class="header-section-number">6</span> Scaffolding with Ragout</h1>
<p>In this chapter, we will use Ragout to scaffold our best de novo assembly. Scaffolding is the process of ordering and orienting contigs from a draft assembly into larger structures, called scaffolds, using reference genomes. This step is crucial for improving the contiguity of the assembly and approaching a chromosome-level assembly.</p>
<p>We will use multiple reference genomes to guide the scaffolding process. Ragout will infer the correct order and orientation of the contigs based on synteny information from these references.</p>
<section id="configuration-and-setup" class="level2" data-number="6.1">
<h2 data-number="6.1" class="anchored" data-anchor-id="configuration-and-setup"><span class="header-section-number">6.1</span> Configuration and Setup</h2>
<section id="environment-variables" class="level3" data-number="6.1.1">
<h3 data-number="6.1.1" class="anchored" data-anchor-id="environment-variables"><span class="header-section-number">6.1.1</span> Environment Variables</h3>
<p>We’ll continue using the established environment variables:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="bu">export</span> <span class="va">SN</span><span class="op">=</span>tuto482</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="bu">export</span> <span class="va">NCPUS</span><span class="op">=</span>12</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="software-dependencies" class="level3" data-number="6.1.2">
<h3 data-number="6.1.2" class="anchored" data-anchor-id="software-dependencies"><span class="header-section-number">6.1.2</span> Software Dependencies</h3>
<p>We will use Ragout for this chapter.</p>
<section id="apptainer" class="level4" data-number="6.1.2.1">
<h4 data-number="6.1.2.1" class="anchored" data-anchor-id="apptainer"><span class="header-section-number">6.1.2.1</span> Apptainer</h4>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="bu">alias</span> ragout=<span class="st">"apptainer run docker://ghcr.io/nexomis/ragout:build_1.0-ragout_2.3-hal_2.3-Sibelia_3.0.7 ragout"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="docker" class="level4" data-number="6.1.2.2">
<h4 data-number="6.1.2.2" class="anchored" data-anchor-id="docker"><span class="header-section-number">6.1.2.2</span> Docker</h4>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="bu">alias</span> ragout=<span class="st">"docker run --rm -u </span><span class="va">$UID</span><span class="st">:</span><span class="va">$GID</span><span class="st"> -v </span><span class="va">$(</span><span class="bu">pwd</span><span class="va">)</span><span class="st">:</span><span class="va">$(</span><span class="bu">pwd</span><span class="va">)</span><span class="st"> -w </span><span class="va">$(</span><span class="bu">pwd</span><span class="va">)</span><span class="st"> ghcr.io/nexomis/ragout:build_1.0-ragout_2.3-hal_2.3-Sibelia_3.0.7 ragout"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="conda" class="level4" data-number="6.1.2.3">
<h4 data-number="6.1.2.3" class="anchored" data-anchor-id="conda"><span class="header-section-number">6.1.2.3</span> Conda</h4>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> create <span class="at">-n</span> ragout <span class="at">-c</span> bioconda ragout</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate ragout</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
</section>
<section id="directory-structure-setup" class="level3" data-number="6.1.3">
<h3 data-number="6.1.3" class="anchored" data-anchor-id="directory-structure-setup"><span class="header-section-number">6.1.3</span> Directory Structure Setup</h3>
<p>Let’s create an organized directory structure for our analysis:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> 06-scaffolding/references</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> 06-scaffolding/ragout_output</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
</section>
<section id="data-preparation" class="level2" data-number="6.2">
<h2 data-number="6.2" class="anchored" data-anchor-id="data-preparation"><span class="header-section-number">6.2</span> Data Preparation</h2>
<section id="download-reference-genomes" class="level3" data-number="6.2.1">
<h3 data-number="6.2.1" class="anchored" data-anchor-id="download-reference-genomes"><span class="header-section-number">6.2.1</span> Download Reference Genomes</h3>
<p>We will download four reference genomes to be used by Ragout.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-P</span> 06-scaffolding/references https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/013/085/055/GCF_013085055.1_ASM1308505v1/GCF_013085055.1_ASM1308505v1_genomic.fna.gz</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-P</span> 06-scaffolding/references https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/048/165/035/GCA_048165035.1_ASM4816503v1/GCA_048165035.1_ASM4816503v1_genomic.fna.gz</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-P</span> 06-scaffolding/references https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/048/164/945/GCA_048164945.1_ASM4816494v1/GCA_048164945.1_ASM4816494v1_genomic.fna.gz</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-P</span> 06-scaffolding/references https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/030/719/095/GCA_030719095.1_ASM3071909v1/GCA_030719095.1_ASM3071909v1_genomic.fna.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Ragout requires the input fasta files to be uncompressed and renamed with simple names.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Decompress and rename reference files</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="fu">gunzip</span> 06-scaffolding/references/<span class="pp">*</span>.gz</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="fu">mv</span> 06-scaffolding/references/GCF_013085055.1_ASM1308505v1_genomic.fna 06-scaffolding/references/Fo47.fasta</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="fu">mv</span> 06-scaffolding/references/GCA_048165035.1_ASM4816503v1_genomic.fna 06-scaffolding/references/ZUM2407.fasta</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="fu">mv</span> 06-scaffolding/references/GCA_048164945.1_ASM4816494v1_genomic.fna 06-scaffolding/references/V032g.fasta</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a><span class="fu">mv</span> 06-scaffolding/references/GCA_030719095.1_ASM3071909v1_genomic.fna 06-scaffolding/references/ME23.fasta</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a><span class="fu">cp</span> 04-assembly/masurca/option-A/flye<span class="pp">*</span>/assembly.fasta 06-scaffolding/references/<span class="va">${SN}</span>.fasta</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="prepare-ragout-recipe-file" class="level3" data-number="6.2.2">
<h3 data-number="6.2.2" class="anchored" data-anchor-id="prepare-ragout-recipe-file"><span class="header-section-number">6.2.2</span> Prepare Ragout Recipe File</h3>
<p>Ragout uses a recipe file to specify the input genomes and parameters. We will create one for our analysis. We will use the best assembly from the previous step (<code>04-assembly/masurca/option-A/flye*/assembly.fasta</code>) as the target assembly.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Create the recipe file</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span> <span class="op">&gt;</span> 06-scaffolding/ragout_recipe.rcp <span class="op">&lt;&lt; EOF</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="st">.references = Fo47,ZUM2407,V032g,ME23</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a><span class="st">.target = </span><span class="va">${SN}</span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a><span class="st">.naming_ref = Fo47</span></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a><span class="st">Fo47.fasta = </span><span class="va">${PWD}</span><span class="st">/06-scaffolding/references/Fo47.fasta</span></span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a><span class="st">ZUM2407.fasta = </span><span class="va">${PWD}</span><span class="st">/06-scaffolding/references/ZUM2407.fasta</span></span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a><span class="st">V032g.fasta = </span><span class="va">${PWD}</span><span class="st">/06-scaffolding/references/V032g.fasta </span></span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a><span class="st">ME23.fasta = </span><span class="va">${PWD}</span><span class="st">/06-scaffolding/references/ME23.fasta</span></span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a><span class="va">${SN}</span><span class="st">.fasta = </span><span class="va">${PWD}</span><span class="st">/06-scaffolding/references/</span><span class="va">${SN}</span><span class="st">.fasta</span></span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a><span class="op">EOF</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
</section>
<section id="ragout-with-sibelia-default-method" class="level2" data-number="6.3">
<h2 data-number="6.3" class="anchored" data-anchor-id="ragout-with-sibelia-default-method"><span class="header-section-number">6.3</span> Ragout with Sibelia (Default Method)</h2>
<p>The default method uses Sibelia for synteny block decomposition. While effective, this can be slow for larger genomes. We’ll run Ragout with the recipe file we created using the <code>--refine</code> option to improve the assembly using the assembly graph.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="ex">ragout</span> 06-scaffolding/ragout_recipe.rcp <span class="dt">\</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">-o</span> 06-scaffolding/ragout_output <span class="dt">\</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">--refine</span> <span class="dt">\</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">--solid-scaffolds</span> <span class="dt">\</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">-t</span> <span class="va">$NCPUS</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><strong>Note</strong>: This approach can be time-consuming for larger genomes. For faster processing, consider using the HAL alignment method described below.</p>
</section>
<section id="ragout-with-cactus-hal-alignment-method" class="level2" data-number="6.4">
<h2 data-number="6.4" class="anchored" data-anchor-id="ragout-with-cactus-hal-alignment-method"><span class="header-section-number">6.4</span> Ragout with Cactus (HAL Alignment Method)</h2>
<p>For larger genomes (&gt;100Mb) or when faster processing is desired, we can use Cactus to create a HAL (Hierarchical Alignment) file, which Ragout can then use for synteny block decomposition. This approach is generally faster and more suitable for complex genomes.</p>
<section id="software-dependencies-for-cactus-and-jolytree" class="level3" data-number="6.4.1">
<h3 data-number="6.4.1" class="anchored" data-anchor-id="software-dependencies-for-cactus-and-jolytree"><span class="header-section-number">6.4.1</span> Software Dependencies for Cactus and JolyTree</h3>
<section id="create-local-conda-environment-for-jolytree" class="level4" data-number="6.4.1.1">
<h4 data-number="6.4.1.1" class="anchored" data-anchor-id="create-local-conda-environment-for-jolytree"><span class="header-section-number">6.4.1.1</span> Create Local Conda Environment for JolyTree</h4>
<p>Since the JolyTree container is missing dependencies, we’ll create a local conda environment:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create local conda environment in hands-on/venv</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> create <span class="at">-p</span> venv <span class="at">-c</span> bioconda <span class="at">-c</span> conda-forge jolytree</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="apptainer-cactus-only" class="level4" data-number="6.4.1.2">
<h4 data-number="6.4.1.2" class="anchored" data-anchor-id="apptainer-cactus-only"><span class="header-section-number">6.4.1.2</span> Apptainer (Cactus only)</h4>
<div class="sourceCode" id="cb11"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="bu">alias</span> cactus=<span class="st">"apptainer run docker://quay.io/comparative-genomics-toolkit/cactus:v2.9.9 cactus"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="docker-cactus-only" class="level4" data-number="6.4.1.3">
<h4 data-number="6.4.1.3" class="anchored" data-anchor-id="docker-cactus-only"><span class="header-section-number">6.4.1.3</span> Docker (Cactus only)</h4>
<div class="sourceCode" id="cb12"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="bu">alias</span> cactus=<span class="st">"docker run --rm -u </span><span class="va">$UID</span><span class="st">:</span><span class="va">$GID</span><span class="st"> -v </span><span class="va">$(</span><span class="bu">pwd</span><span class="va">)</span><span class="st">:</span><span class="va">$(</span><span class="bu">pwd</span><span class="va">)</span><span class="st"> -w </span><span class="va">$(</span><span class="bu">pwd</span><span class="va">)</span><span class="st"> quay.io/comparative-genomics-toolkit/cactus:v2.9.9 cactus"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><strong>Note</strong>: JolyTree.sh will be available directly from the activated conda environment.</p>
</section>
</section>
<section id="generate-phylogenetic-tree-with-jolytree" class="level3" data-number="6.4.2">
<h3 data-number="6.4.2" class="anchored" data-anchor-id="generate-phylogenetic-tree-with-jolytree"><span class="header-section-number">6.4.2</span> Generate Phylogenetic Tree with JolyTree</h3>
<p>Before creating the Cactus sequence file, we’ll use JolyTree to generate an accurate phylogenetic tree from our reference genomes and target assembly. JolyTree is a fast alignment-free tool that can infer distance-based phylogenetic trees from genome assemblies.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Run JolyTree to generate phylogenetic tree</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate ./venv</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="ex">JolyTree.sh</span> <span class="at">-i</span> 06-scaffolding/references <span class="at">-b</span> 06-scaffolding/jolytree <span class="at">-t</span> <span class="va">$NCPUS</span></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> deactivate</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><strong>JolyTree parameters explained:</strong></p>
<ul>
<li><code>-i</code>: Directory containing FASTA files (all genomes)</li>
<li><code>-b</code>: Basename for output files</li>
<li><code>-t</code>: Number of threads to use</li>
</ul>
<p>The generated tree file <code>06-scaffolding/jolytree.nwk</code> will contain the inferred phylogenetic relationships between all genomes.</p>
<p><strong>Note on Alternative Tree Construction Methods:</strong></p>
<p>JolyTree uses a sound methodology (Mash for distance estimation followed by tree construction algorithms), but its implementation lacks robustness and reproducibility. For more rigorous phylogenetic analysis, you can manually construct pairwise distance matrices based on k-mer frequencies using tools like:</p>
<ul>
<li><strong>KAT</strong>: Can generate k-mer frequency histograms and compare k-mer profiles between genomes</li>
<li><strong>Mash</strong>: Estimates genome-to-genome distances using MinHash sketches of k-mer sets</li>
</ul>
<p>These distance matrices can then be used with phylogenetic reconstruction methods such as:</p>
<ul>
<li><strong>UPGMA</strong> (Unweighted Pair Group Method with Arithmetic Mean)</li>
<li><strong>Neighbor-joining</strong></li>
</ul>
<p>This manual approach provides better control over parameters and more reproducible results, especially for closely related genomes or when branch length accuracy is important for downstream analyses.</p>
</section>
<section id="manual-tree-editing-for-cactus-compatibility" class="level3" data-number="6.4.3">
<h3 data-number="6.4.3" class="anchored" data-anchor-id="manual-tree-editing-for-cactus-compatibility"><span class="header-section-number">6.4.3</span> Manual Tree Editing for Cactus Compatibility</h3>
<p><strong>Important</strong>: The JolyTree output may need manual editing to be compatible with Cactus requirements:</p>
<ol type="1">
<li><strong>Remove branch lengths</strong>: Cactus can work without branch lengths.</li>
<li><strong>Convert multifurcations to binary tree</strong>: Cactus requires strictly binary trees (each internal node has exactly 2 children). If JolyTree produces any 3-way or higher splits, these need to be resolved into binary splits.</li>
</ol>
<p>To edit the tree <code>06-scaffolding/jolytree.nwk</code> manually:</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="kw">(</span><span class="ex">tuto482:0.1,Fo47:0.2,</span><span class="er">(</span><span class="ex">ZUM2407:0.05,V032g:0.03,ME23:0.08</span><span class="kw">)</span><span class="ex">:0.15</span><span class="kw">);</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>into</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="kw">((</span><span class="ex">tuto482,Fo47</span><span class="kw">)</span><span class="ex">,</span><span class="er">((</span><span class="va">ZUM2407</span><span class="kw">,</span><span class="va">V032g</span><span class="kw">)</span><span class="ex">,ME23</span><span class="kw">));</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><strong>Tree requirements for Cactus:</strong></p>
<ul>
<li>Must be in Newick format</li>
<li>Must be strictly binary (no multifurcations)</li>
<li>All leaf nodes must correspond to genome names</li>
<li>Branch lengths are ignored</li>
</ul>
</section>
<section id="create-cactus-sequence-file" class="level3" data-number="6.4.4">
<h3 data-number="6.4.4" class="anchored" data-anchor-id="create-cactus-sequence-file"><span class="header-section-number">6.4.4</span> Create Cactus Sequence File</h3>
<p>Now we’ll create a sequence file for Cactus using the phylogenetic tree generated by JolyTree:</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create directory for Cactus workflow</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> 06-scaffolding/cactus</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span> <span class="op">&gt;</span> 06-scaffolding/cactus/genomes.txt <span class="op">&lt;&lt; EOF</span></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a><span class="st"># Phylogenetic tree generated by JolyTree</span></span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a><span class="st">((ME23,(ZUM2407,(V032g,Fo47))),tuto482);</span></span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a><span class="st"># Genome paths (must be absolute paths)</span></span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a><span class="va">${SN}</span><span class="st"> </span><span class="va">${PWD}</span><span class="st">/06-scaffolding/references/</span><span class="va">${SN}</span><span class="st">.fasta</span></span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a><span class="st">Fo47 </span><span class="va">${PWD}</span><span class="st">/06-scaffolding/references/Fo47.fasta</span></span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a><span class="st">ZUM2407 </span><span class="va">${PWD}</span><span class="st">/06-scaffolding/references/ZUM2407.fasta</span></span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true" tabindex="-1"></a><span class="st">V032g </span><span class="va">${PWD}</span><span class="st">/06-scaffolding/references/V032g.fasta</span></span>
<span id="cb16-13"><a href="#cb16-13" aria-hidden="true" tabindex="-1"></a><span class="st">ME23 </span><span class="va">${PWD}</span><span class="st">/06-scaffolding/references/ME23.fasta</span></span>
<span id="cb16-14"><a href="#cb16-14" aria-hidden="true" tabindex="-1"></a><span class="op">EOF</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="run-cactus-to-generate-hal-alignment" class="level3" data-number="6.4.5">
<h3 data-number="6.4.5" class="anchored" data-anchor-id="run-cactus-to-generate-hal-alignment"><span class="header-section-number">6.4.5</span> Run Cactus to Generate HAL Alignment</h3>
<div class="sourceCode" id="cb17"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Run Cactus to create HAL alignment</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="ex">cactus</span> 06-scaffolding/cactus/js <span class="dt">\</span></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>  06-scaffolding/cactus/genomes.txt <span class="dt">\</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>  06-scaffolding/cactus/alignment.hal <span class="dt">\</span></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">--maxCores</span> <span class="va">$NCPUS</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><strong>Cactus parameters explained:</strong></p>
<ul>
<li><code>js</code>: Job store directory for intermediate files</li>
<li><code>genomes.txt</code>: Sequence file with tree and genome paths</li>
<li><code>alignment.hal</code>: Output HAL alignment file</li>
<li><code>--maxCores</code>: Maximum number of CPU cores to use</li>
</ul>
</section>
<section id="convert-hal-to-maf-format" class="level3" data-number="6.4.6">
<h3 data-number="6.4.6" class="anchored" data-anchor-id="convert-hal-to-maf-format"><span class="header-section-number">6.4.6</span> Convert HAL to MAF Format</h3>
<p>After generating the HAL alignment, we need to convert it to MAF format for use with Ragout:</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Set up cactus-hal2maf alias</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="bu">alias</span> cactus-hal2maf=<span class="st">"apptainer run docker://quay.io/comparative-genomics-toolkit/cactus:v2.9.9 cactus-hal2maf"</span></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert HAL to MAF format</span></span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a><span class="ex">cactus-hal2maf</span> 06-scaffolding/cactus/js-hal2maf <span class="dt">\</span></span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>  06-scaffolding/cactus/alignment.hal <span class="dt">\</span></span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a>  06-scaffolding/cactus/alignment.maf <span class="dt">\</span></span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">--refGenome</span> Fo47 <span class="dt">\</span></span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">--filterGapCausingDupes</span> <span class="dt">\</span></span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">--noAncestors</span> <span class="dt">\</span></span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">--dupeMode</span> single <span class="dt">\</span></span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">--chunkSize</span> 500000</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><strong>cactus-hal2maf parameters explained:</strong></p>
<ul>
<li><code>--refGenome</code>: Reference genome for MAF projection (Fo47 for consistent chromosome naming)</li>
<li><code>--chunkSize</code>: Size of chunks for parallel processing</li>
</ul>
</section>
<section id="create-ragout-recipe-for-maf-method" class="level3" data-number="6.4.7">
<h3 data-number="6.4.7" class="anchored" data-anchor-id="create-ragout-recipe-for-maf-method"><span class="header-section-number">6.4.7</span> Create Ragout Recipe for MAF Method</h3>
<p>Now create a recipe file that uses the MAF alignment:</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create MAF-based recipe file</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span> <span class="op">&gt;</span> 06-scaffolding/ragout_maf_recipe.rcp <span class="op">&lt;&lt; EOF</span></span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a><span class="st">.references = Fo47,ZUM2407,V032g,ME23</span></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a><span class="st">.target = </span><span class="va">${SN}</span></span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a><span class="st">.naming_ref = Fo47</span></span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a><span class="st">.maf = </span><span class="va">${PWD}</span><span class="st">/06-scaffolding/cactus/alignment.maf</span></span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a><span class="st"># Individual FASTA paths are required for MAF backend</span></span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a><span class="st">Fo47.fasta = </span><span class="va">${PWD}</span><span class="st">/06-scaffolding/references/Fo47.fasta</span></span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a><span class="st">ZUM2407.fasta = </span><span class="va">${PWD}</span><span class="st">/06-scaffolding/references/ZUM2407.fasta</span></span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a><span class="st">V032g.fasta = </span><span class="va">${PWD}</span><span class="st">/06-scaffolding/references/V032g.fasta</span></span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a><span class="st">ME23.fasta = </span><span class="va">${PWD}</span><span class="st">/06-scaffolding/references/ME23.fasta</span></span>
<span id="cb19-13"><a href="#cb19-13" aria-hidden="true" tabindex="-1"></a><span class="va">${SN}</span><span class="st">.fasta = </span><span class="va">${PWD}</span><span class="st">/06-scaffolding/references/</span><span class="va">${SN}</span><span class="st">.fasta</span></span>
<span id="cb19-14"><a href="#cb19-14" aria-hidden="true" tabindex="-1"></a><span class="op">EOF</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="run-ragout-with-maf-alignment" class="level3" data-number="6.4.8">
<h3 data-number="6.4.8" class="anchored" data-anchor-id="run-ragout-with-maf-alignment"><span class="header-section-number">6.4.8</span> Run Ragout with MAF Alignment</h3>
<div class="sourceCode" id="cb20"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="ex">ragout</span> 06-scaffolding/ragout_maf_recipe.rcp <span class="dt">\</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">-o</span> 06-scaffolding/ragout_maf_output <span class="dt">\</span></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">--solid-scaffolds</span> <span class="dt">\</span></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">-s</span> maf <span class="dt">\</span></span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">-t</span> <span class="va">$NCPUS</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><strong>MAF method parameters:</strong></p>
<ul>
<li><code>-s maf</code>: Specifies MAF as the synteny backend</li>
<li>Individual FASTA paths must be provided alongside the MAF file</li>
</ul>
</section>
</section>
<section id="method-comparison" class="level2" data-number="6.5">
<h2 data-number="6.5" class="anchored" data-anchor-id="method-comparison"><span class="header-section-number">6.5</span> Method Comparison</h2>
<table class="caption-top table">
<colgroup>
<col style="width: 24%">
<col style="width: 24%">
<col style="width: 26%">
<col style="width: 24%">
</colgroup>
<thead>
<tr class="header">
<th>Method</th>
<th>Processing Time</th>
<th>Genome Size Suitability</th>
<th>Memory Usage</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><strong>Sibelia</strong></td>
<td>Slower</td>
<td>&lt;100Mb (bacterial, small eukaryotes)</td>
<td>Moderate</td>
</tr>
<tr class="even">
<td><strong>HAL/Cactus</strong></td>
<td>Faster</td>
<td>&gt;100Mb (large eukaryotes)</td>
<td>Higher initial, but more efficient overall</td>
</tr>
</tbody>
</table>
<p><strong>Recommendations:</strong></p>
<ul>
<li>Use <strong>Sibelia method</strong> (default) for bacterial genomes or small eukaryotic genomes</li>
<li>Use <strong>HAL/Cactus method</strong> for large eukaryotic genomes or when faster processing is needed</li>
</ul>
</section>
<section id="rna-seq-based-scaffolding-with-agouti" class="level2" data-number="6.6">
<h2 data-number="6.6" class="anchored" data-anchor-id="rna-seq-based-scaffolding-with-agouti"><span class="header-section-number">6.6</span> RNA-Seq Based Scaffolding with Agouti</h2>
<p>(This part will be seen after annotation tutorial)</p>
</section>
<section id="expected-outcomes" class="level2" data-number="6.7">
<h2 data-number="6.7" class="anchored" data-anchor-id="expected-outcomes"><span class="header-section-number">6.7</span> Expected Outcomes</h2>
<p>After Ragout finishes, the output directory will contain the scaffolded assembly. The main output files are:</p>
<section id="for-sibelia-method-06-scaffoldingragout_output" class="level3" data-number="6.7.1">
<h3 data-number="6.7.1" class="anchored" data-anchor-id="for-sibelia-method-06-scaffoldingragout_output"><span class="header-section-number">6.7.1</span> For Sibelia Method (<code>06-scaffolding/ragout_output/</code>):</h3>
<ul>
<li><code>${SN}_scaffolds.fasta</code>: The final scaffolded assembly</li>
<li><code>${SN}_unplaced.fasta</code>: Input contigs that could not be placed in the scaffolds</li>
<li><code>${SN}_scaffolds.agp</code>: The structure of the scaffolds in AGP format</li>
</ul>
</section>
<section id="for-cactus-method-06-scaffoldingragout_maf_output" class="level3" data-number="6.7.2">
<h3 data-number="6.7.2" class="anchored" data-anchor-id="for-cactus-method-06-scaffoldingragout_maf_output"><span class="header-section-number">6.7.2</span> For Cactus Method (<code>06-scaffolding/ragout_maf_output/</code>):</h3>
<ul>
<li><code>${SN}_scaffolds.fasta</code>: The final scaffolded assembly</li>
<li><code>${SN}_unplaced.fasta</code>: Input contigs that could not be placed in the scaffolds</li>
<li><code>${SN}_scaffolds.agp</code>: The structure of the scaffolds in AGP format</li>
</ul>
</section>
<section id="additional-cactus-outputs-06-scaffoldingcactus" class="level3" data-number="6.7.3">
<h3 data-number="6.7.3" class="anchored" data-anchor-id="additional-cactus-outputs-06-scaffoldingcactus"><span class="header-section-number">6.7.3</span> Additional Cactus Outputs (<code>06-scaffolding/cactus/</code>):</h3>
<ul>
<li><code>alignment.hal</code>: Multiple genome alignment in HAL format</li>
<li><code>alignment.maf</code>: Multiple genome alignment in HAL format</li>
</ul>
<p>You can then proceed to evaluate the quality of the new scaffolded assembly using QUAST and BUSCO, as described in the previous chapter. Compare the results from both methods to determine which approach works best for your specific dataset.</p>


</section>
</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("https:\/\/nexomis\.github\.io\/tuto-euk-de-novo-genome\/");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




</body></html>